[FIX] sale_order_revision #184

thomaspaulb opened the pull request:
Fix: set new order lines to draft and old ones to cancel.

Rationale: the way this module creates a new revision is that the old, cancelled revision is renamed to the new revision, then a copy is created and named after the old revision. As a result the lines of the old revision (copy) have state "draft" (the default) and the lines of the new revision have state "cancel". It should be the other way around. So, this fix switches the states around.

Commit:
Issue #184 fix
Set new order lines to draft and old ones to cancel

flake8

flake8

eLBati commented:
Thanks
Please see travis

```
./sale_order_revision/model/sale_order.py:69:68: W291 trailing whitespace
./sale_order_revision/model/sale_order.py:70:13: E128 continuation line under-indented for visual indent
./sale_order_revision/model/sale_order.py:70:59: W291 trailing whitespace
./sale_order_revision/model/sale_order.py:74:13: E128 continuation line under-indented for visual indent
```

Also, could you extend the automated test adding the failing test case?

Commit:
Unit test to assert right state of order lines

flake8

PEP

There was another problem, the old order lines became the new order lines and vice versa, but the old procurement order referred to the old order line and thus now to the new order line, so that when creating a shipment, the old procurement order was used containing the wrong lines. Long story short, instead of also switching the procurement order around, a better strategy is to just swap the lines of the old and new order completely.

Style improvements

thomaspaulb commented:
Ok I think it is better now, how can I run the unit test to check?

petrus-v commented:
Hi @antiflu,

To run yaml test locally you can use `--test-enable` option of your odoo server script.
:+1: for the fix

thomaspaulb commented:
So yea we found another bug with the procurement/shipping, also to do with order lines being swapped... so I changed the fix to really swap the lines between orders, not only toggle the line state. Should be fine now, but we're going to production-test it again. 

@petrus-v I managed to successfully run the tests, so they are OK as well. ("tested" the tests...)

pedrobaeza commented on sale_order_revision/model/sale_order.py:
```diff
@@ -64,6 +69,20 @@ def copy_quotation(self):
+
+        # put existing procurement group id on old order
+        old_revision.write({'procurement_group_id': procurement_group_id})
+        self.write({'procurement_group_id': None})
```
Why do you write this 2 times?

thomaspaulb replied:
How to write it in one?

pedrobaeza replied:
I mean that you have the same sentence in line 60 and in line 75

pedrobaeza commented on sale_order_revision/model/sale_order.py:
```diff
@@ -64,6 +69,20 @@ def copy_quotation(self):
+        self.write({'procurement_group_id': None})
+
+        # swap order lines of old and new order
+        old_line_ids = [line.id for line in old_revision.order_line]
```
Why don't `old_revision.order_line.ids`?

pedrobaeza commented on sale_order_revision/model/sale_order.py:
```diff
@@ -64,6 +69,20 @@ def copy_quotation(self):
+
+        # swap order lines of old and new order
+        old_line_ids = [line.id for line in old_revision.order_line]
+        new_line_ids = [line.id for line in self.order_line]
```
The same: `self.order_line.ids`

pedrobaeza commented on sale_order_revision/model/sale_order.py:
```diff
@@ -64,6 +69,20 @@ def copy_quotation(self):
+        # swap order lines of old and new order
+        old_line_ids = [line.id for line in old_revision.order_line]
+        new_line_ids = [line.id for line in self.order_line]
+        so_line = self.pool.get('sale.order.line')
```
Why not new API?

pedrobaeza commented:
Can you please check also the procurement group in the tests?

Commit:
Procurement test, is however dependent on sale_stock

Remove sale_stock dependency and the tests that rely on it

No need to rebrowse

thomaspaulb commented:
Implemented your suggestions in the first commit.

Second commit contains the procurement group test... however:
- procurement group creation requires order_policy = picking
- order_policy = picking requires sale_stock to be listed as dependency
- rest of the module does not necessarily have to depend on sale_stock

This is a logical conundrum that I have run into before, maybe you know the answer - how to make use of a module _when it's there_, but otherwise leave it be? Or in this case, how to extend a test to test cases where there is dependency on a certain module, but not make the whole module dependent? Only answer I have found to that so far is to make a new submodule `sale_order_revision_stock` that contains the stock dependencies...

nhomar commented:
@antiflu 

Just add to your python class for test:

``` python
@openerp.tests.common.at_install(False)
@openerp.tests.common.post_install(True)
```

On this way your module will be tested when everything finished AFAICU.

But frequently if it is such case is because you have a design issue that need some other workaround more efective (may be a glue module which test the case dependant).

regards.

eLBati commented on sale_order_revision/model/sale_order.py:
```diff
@@ -64,6 +69,17 @@ def copy_quotation(self):
+        old_lines = so_line.browse(old_revision.order_line.ids)
+        new_lines = so_line.browse(self.order_line.ids)
+        old_lines.write({'order_id': self.id})
+        new_lines.write({'order_id': old_revision.id})
```
You don't need `old_lines` and `new_lines` variables. You can call `old_revision.order_line.write` and `self.order_line.write`

thomaspaulb replied:
I tested your shorter version; indeed it goes wrong, the old revision will contain both revision's lines and the new revision contains nothing.

eLBati commented:
@antiflu @pedrobaeza in this case, I would just remove lines
https://github.com/OCA/sale-workflow/pull/186/files#diff-7afbf0b5a2943e16e7a3935595fdf635R8
and
https://github.com/OCA/sale-workflow/pull/186/files#diff-7afbf0b5a2943e16e7a3935595fdf635R35
from the test

I would not make this module depending on sale_stock just for tests purpose, as its functionalities can also be used without sale_stock.
Possibly, we could create an hidden module `sale_order_revision_stock_test`, containing the missing test only

thomaspaulb commented on sale_order_revision/model/sale_order.py:
```diff
@@ -64,6 +69,17 @@ def copy_quotation(self):
+        old_lines = so_line.browse(old_revision.order_line.ids)
+        new_lines = so_line.browse(self.order_line.ids)
+        old_lines.write({'order_id': self.id})
+        new_lines.write({'order_id': old_revision.id})
```
I wanted to do that, but I was afraid that after the old_revision.order_line.write action, the new lines will now belong to the old revision and thus self.order_line.ids will contain no items and the second write action will be flawed.

eLBati replied:
> I wanted to do that, but I was afraid that after the
> old_revision.order_line.write action, the new lines will now belong to the
> old revision and thus self.order_line.ids will contain no items and the
> second write action will be flawed.
> 
> Did this actually happen?

thomaspaulb commented:
Sounds interesting. Can a hidden module be in a subfolder of the normal module?

eLBati commented:
On 13 August 2015 at 10:55, Tom notifications@github.com wrote:

> Sounds interesting. Can a hidden module be in a subfolder of the normal
> module?
> 
> No. It would just have Hidden category.
> See for instance web_calendar/**openerp**.py

pablocm-aserti commented:
Maybe I'm missing something, but sometimes it is neccessary to take 2 steps back to see the whole picture.
What is the reason to do a duplicate, disguise it as the old SO and rename the original one to look like a new revision?
Wouldn't it be easier and better for maintainability to just use the duplicate as the new revision with fresh new lines?

As I see it, the cancel status problem and the procurement management needed to make it work come from that decision.
Furthermore, this could lead to data inconsistencies for third-party modules that add objects pointing to those lines, since they are now in another SO.

thomaspaulb commented:
I think that before, the module worked like you described, but then someone
changed it around, and the problems that I fixed arose.

The reason I did not change it back is that I thought someone might have
had a reason for doing it this way. A reason I could think of is that when
there is somewhere in some module a reference to an older revision, and
there comes a new revision, in this way the references will automatically
move to the new revision. Instead, if it is done your way, the whole
database needs to be scanned for references to the old revision and updated
to the id of the new revision. Which would not be desirable.

Does that make sense to you?

On Thu, Oct 22, 2015 at 11:25 AM, Pablo C.M. notifications@github.com
wrote:

> Maybe I'm missing something, but sometimes it is neccessary to take 2
> steps back to see the whole picture.
> What is the reason to do a duplicate, disguise it as the old SO and rename
> the original one to look like a new revision?
> Wouldn't it be easier and better for maintainability to just use the
> duplicate as the new revision with fresh new lines?
> 
> As I see it, the cancel status problem and the procurement management
> needed to make it work come from that decision.
> Furthermore, this could lead to data inconsistencies for third-party
> modules that add objects pointing to those lines, since they are now in
> another SO.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/sale-workflow/pull/186#issuecomment-150159763.

pablocm-aserti commented:
> A reason I could think of is that when there is somewhere in some module a reference to an older revision

But third-party modules -including procurement- are not aware of any revision system, just the standard copy feature. They can decide what to do with the the fields/related objects they add -whether carry them over to the new duplicate or not, do a deep copy, etc-. If you do a duplicate and then switch the lines, they know nothing of it and then the responsibility for fixing consistency relies on this module -eg. procurements- or "link modules" to fix things when you use revisions.

> Instead, if it is done your way, the whole database needs to be scanned for references to the old revision and updated to the id of the new revision. Which would not be desirable.

Probably the most desired behaviour is not to change references but let them be, and then let the brand new copy follow its workflow from the beginning, regenerating all those related objects: new procurement group, procurements and so on

pedrobaeza commented:
One of the reason to make a copy, but preserve the current record as the "new revision" is for not losing communication history - aka chatter -, and also not losing possible attachments, links, and so on.

For me is totally right and it must be this way, so let's fix this one and continue.

thomaspaulb commented:
Ok so what remains to be done now?

On Thu, Oct 22, 2015 at 1:33 PM, Pedro M. Baeza notifications@github.com
wrote:

> One of the reason to make a copy, but preserve the current record as the
> "new revision" is for not losing communication history - aka chatter -, and
> also not losing possible attachments, links, and so on.
> 
> For me is totally right and it must be this way, so let's fix this one and
> continue.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/sale-workflow/pull/186#issuecomment-150191005.

pedrobaeza commented:
There's still a comment unattended

pablocm-aserti commented:
Thanks, that's the "something" I was missing.
Maybe it would still be better to carry over the chat history to the new copy because those objects don't have any business logic, but if you say so then it is fine.
If it is useful, since the pending comment is just for the test part, I've reviewed and functionally tested the fix and everything seems OK
Thanks for the fix @antiflu !

eLBati commented:
@antiflu please see
https://github.com/OCA/sale-workflow/pull/186#discussion_r36955270
and
https://github.com/OCA/sale-workflow/pull/186#issuecomment-130573800

pedrobaeza commented on sale_order_revision/model/sale_order.py:
```diff
@@ -64,6 +69,17 @@ def copy_quotation(self):
         old_revision.message_post(body=msg)
+
+        # set stored procurement group id on old order
+        old_revision.write({'procurement_group_id': procurement_group_id})
```
I don't think we should switch procurement group. It's good to have the procurement group set on the newest revision, not in the old one.

thomaspaulb replied:
Good point, have fixed that.

pedrobaeza commented on sale_order_revision/model/sale_order.py:
```diff
@@ -64,6 +69,17 @@ def copy_quotation(self):
+
+        # swap order lines of old and new order
+        so_line = self.env['sale.order.line']
+        old_lines = so_line.browse(old_revision.order_line.ids)
```
You don't need to rebrowse, just assign the recordset: `old_lines = old_revision.order_line`

thomaspaulb replied:
True, have done that now.

pedrobaeza commented:
Any update on this?

thomaspaulb commented:
Sorry guys I am extremely busy at the moment.. it's on my todolist tho

On Tue, Dec 8, 2015 at 10:02 AM, Pedro M. Baeza notifications@github.com
wrote:

> Any update on this?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/sale-workflow/pull/186#issuecomment-162821663.

pedrobaeza commented:
Is this still aplying?

pedrobaeza commented:
It seems this is not going to be finished, so I close.

thomaspaulb commented:
Pedro, I still would like to finish this. Can we reopen. I have addressed the remaining issues.

Commit:
Dont switch procurement group, keep it on the newest revision

pedrobaeza commented:
The branch still has conflicts.

pedrobaeza commented:
Alternative patch in #587
